 Answering Business Questions Using SQL

1. OVERVIEW OF THE DATA

In this project we are going to answer business questions related to Chinook Record Store using SQL. Chinook database is provided as a SQLite database file. Down below we impliment the code which connects the Jupyter Notebook to the Chinook databse.

In [10]:
%%capture
%load_ext sql
%sql sqlite:///chinook.db

'Connected: None@chinook.db'

In [11]:
#We first need to familiarize with the tables and views in the database

In [12]:
%%sql
SELECT 
      name,
      type
FROM sqlite_master
WHERE type IN ("table","view");

Done.


[('album', 'table'),
 ('artist', 'table'),
 ('customer', 'table'),
 ('employee', 'table'),
 ('genre', 'table'),
 ('invoice', 'table'),
 ('invoice_line', 'table'),
 ('media_type', 'table'),
 ('playlist', 'table'),
 ('playlist_track', 'table'),
 ('track', 'table')]

2. SELECTING ALBUMS TO PURCHASE

The Chinook record store has just signed a deal with a new record label, and we have been tasked with selecting the first three albums that will be added to the store, from a list of four.
All four albums are by artists that don't have any tracks in the store right now - we have the artist names, and the genre of music they produce:


Artist Name	              Genre
Regal	                  Hip-Hop
Red Tone	              Punk
Meteor and the Girls	  Pop
Slim Jim Bites	          Blues

The record label specializes in artists from the USA, and they have given Chinook some money to advertise the new albums in the USA, so we're interested in finding out which genres sell the best in the USA

we need to write a query to find out which genres sell the most tracks in the USA. We will write up a summary of the findings, and make a recommendation for the three artists whose albums we should purchase for the store.

In [13]:
%%sql

     WITH t1 AS (SELECT iv.billing_country,
                           il.quantity, 
                           gr.name as genre_name
                      FROM invoice iv
                      JOIN invoice_line il
                        ON iv.invoice_id = il.invoice_id
                      JOIN track tr
                        ON tr.track_id = il.track_id
                      JOIN genre gr
                        ON gr.genre_id = tr.genre_id
                     WHERE iv.billing_country = 'USA'
                   ),

            t2 AS (SELECT genre_name,
                            COUNT(*) AS num_purchases
                       FROM t1
                      GROUP BY 1
                      ORDER BY 2 DESC
                    ) 

        
        SELECT *, 
                ROUND(num_purchases * 100 / 
                      CAST(
                           (SELECT SUM(num_purchases) FROM t2) 
                            AS Float
                          ), 1
                     ) AS percentage_sold
          FROM t2;

             
      
      
      

Done.


[('Rock', 561, 53.4),
 ('Alternative & Punk', 130, 12.4),
 ('Metal', 124, 11.8),
 ('R&B/Soul', 53, 5.0),
 ('Blues', 36, 3.4),
 ('Alternative', 35, 3.3),
 ('Latin', 22, 2.1),
 ('Pop', 22, 2.1),
 ('Hip Hop/Rap', 20, 1.9),
 ('Jazz', 14, 1.3),
 ('Easy Listening', 13, 1.2),
 ('Reggae', 6, 0.6),
 ('Electronica/Dance', 5, 0.5),
 ('Classical', 4, 0.4),
 ('Heavy Metal', 3, 0.3),
 ('Soundtrack', 2, 0.2),
 ('TV Shows', 1, 0.1)]

From the above list it is clear that Rock sells best in USA, i.e. 53.38% (more than half). But we do not have it in the record label.

Below we list all the four genres in the record label according to their sales percentage (highest to lowest) in USA.

Highest of these is Alternate & Punk. It has a sales of 12.37%. And Punk is in the record label.

Next highest is Blues with the sales of 3.43%.
Third highest is Pop with the sales of 2.09%.
Last is Hip-Hop/Rap with only 1.9% sales.
Based on our observation of the sales percentage, we recommend these three artists:-

1.Red Tone

2.Slim Jim Bites

3.Meteor and the Girls

3. Analyzing Employee Sales Performance

Each customer for the Chinook store gets assigned to a sales support agent within the company when they first make a purchase. We have been asked to analyze the purchases of customers belonging to each employee to see if any sales support agent is performing either better or worse than the others.

We might like to consider whether any extra columns from the employee table explain any variance you see, or whether the variance might instead be indicative of employee performance.

In [14]:
%%sql
SELECT e.first_name || " " || e.last_name Employee_Name,
       e.title Title,
       e.hire_date Hire_Date,
       e.birthdate Birthdate,
       max(invoice_date) latest_invoice_date,
       ROUND(SUM(i.total), 2) Total_Sales_$,
       ROUND(SUM(i.total) * 100 /(SELECT SUM(total)
                                    FROM invoice), 2) Sales_Perc
       
  FROM employee e

LEFT JOIN customer c ON c.support_rep_id = e.employee_id
LEFT JOIN invoice i ON i.customer_id = c.customer_id

WHERE Title = "Sales Support Agent"
GROUP BY 1
ORDER BY 6 DESC;


Done.


[('Jane Peacock', 'Sales Support Agent', '2017-04-01 00:00:00', '1973-08-29 00:00:00', '2020-12-30 00:00:00', 1731.51, 36.77),
 ('Margaret Park', 'Sales Support Agent', '2017-05-03 00:00:00', '1947-09-19 00:00:00', '2020-12-29 00:00:00', 1584.0, 33.63),
 ('Steve Johnson', 'Sales Support Agent', '2017-10-17 00:00:00', '1965-03-03 00:00:00', '2020-12-20 00:00:00', 1393.92, 29.6)]

From the above table we see that Jane Peacock holds the highest sales amount of ~ $1731, followed by Margaret Park with ~ $1584. Last is Steve Johnson with ~ $1394 sales amount.

If we look at their hire_date, Jane Peacock joined first i.e. on 4th April 2017. Margaret Park joined in a month's time i.e. on 3rd May 2017. 

Where as Steve Johnson joined a few month's later, i.e. on 17th October 2017. If we look at the latest invoice date of all the three, there is not much difference.

We observe that both Jane Peacock and Steve Johnson hold better sales value compared to Margaret Park, evethough Jane Peacock joined a month earlier and Steve Johnson joined five months later than her. 

If we Look at the birthdate, both Jane Peacock and Steve Johnson are younger to Margaret Park and Jane Peacock is younger to Steve Johnson. 

From this we perceive that the age can also have an effect on the performance of the employees. i.e. younger employees perform better than older employees.

4. Analyzing Sales by Country

Our next task is to analyze the sales data for customers from each different country. 

In particular, We have been directed to calculate data, for each country, on the:

Because there are a number of countries with only one customer, We should group these customers as "Other" in your analysis. 



We are going to write a query that collates data on purchases from different countries.
Where a country has only one customer, collect them into an "Other" group.

The results should be sorted by the total sales from highest to lowest, with the "Other" group at the very bottom.

For each country, include:
total number of customers
total value of sales
average value of sales per customer
average order value

In [15]:
%%sql
WITH num_of_cstmrs AS
(SELECT country Country,
       count(first_name) No_of_customers       
  FROM customer 
GROUP BY 1
ORDER BY 2 DESC)  

SELECT * FROM num_of_cstmrs;

Done.


[('USA', 13),
 ('Canada', 8),
 ('Brazil', 5),
 ('France', 5),
 ('Germany', 4),
 ('United Kingdom', 3),
 ('Czech Republic', 2),
 ('India', 2),
 ('Portugal', 2),
 ('Argentina', 1),
 ('Australia', 1),
 ('Austria', 1),
 ('Belgium', 1),
 ('Chile', 1),
 ('Denmark', 1),
 ('Finland', 1),
 ('Hungary', 1),
 ('Ireland', 1),
 ('Italy', 1),
 ('Netherlands', 1),
 ('Norway', 1),
 ('Poland', 1),
 ('Spain', 1),
 ('Sweden', 1)]

First we will calculate the total value of sales, average value of sales per customer and average order value for all countries. 

Then we can seperate the countries with only one customer, aggregate their values and rename them as Others. Later we can club the countries with more than one customer and countries with one customer into one table

In [ ]:
%%sql
WITH tot_sales_val AS
(SELECT c.country Country,
       ROUND(SUM(i.total), 2) Tot_Value_Sales,
       ROUND(AVG(i.total), 2) Avg_Order_Value,
       ROUND(SUM(i.total)/(SELECT COUNT(DISTINCT(customer_id)) FROM invoice), 2) Avg_Sales_Value_per_Customer
 
 FROM customer c
INNER JOIN invoice i ON i.customer_id = c.customer_id
GROUP BY 1
ORDER BY 2 DESC),

sales_data_1 AS
(SELECT noc.Country,
       noc.No_of_customers,
       tsv.Tot_Value_Sales,
       tsv.Avg_Sales_Value_per_Customer,
       tsv.Avg_Order_Value
  FROM Num_of_cstmrs noc  
INNER JOIN tot_sales_val tsv ON tsv.Country=noc.Country
 WHERE noc.No_of_customers > 1),

sales_data_2 AS
(SELECT CASE
       WHEN noc.No_of_customers=1 THEN "Others"
       END AS
       Country,
       noc.No_of_customers,
       ROUND(AVG(tsv.Tot_Value_Sales),2) Tot_Value_Sales,
       ROUND(AVG(tsv.Avg_Sales_Value_per_Customer),2) AVG_Sales_Value_per_Customer,
       ROUND(AVG(tsv.Avg_Order_Value),2) AVG_Order_Value
  FROM Num_of_customers noc  
INNER JOIN tot_sales_val tsv ON tsv.Country=noc.Country
 WHERE noc.No_of_customers = 1),

tot_sales_data AS
(SELECT * 
FROM sales_data_1
UNION
SELECT * 
FROM sales_data_2)

SELECT *
  FROM tot_sales_data
ORDER BY 3 DESC;

Above table is in the descending order of total value of sales. And Others (countries with only one customer) is at the bottom of the table.

Here we observe that USA is on top of the list with 13 customers and with ~ $1040 total value of sales. Next is CANADA with 8 customers and ~ $535 total value of sales.

5. Albums vs Individual Tracks

The Chinook store is setup in a way that allows customer to make purchases in one of the two ways:

purchase a whole album
purchase a collection of one or more individual tracks.
The store does not let customers purchase a whole album, and then add individual tracks to that same purchase (unless they do that by choosing each track manually). 

When customers purchase albums they are charged the same price as if they had purchased each of those tracks separately.

Management are currently considering changing their purchasing strategy to save money. 

The strategy they are considering is to purchase only the most popular tracks from each album from record companies, instead of purchasing every track from an album.

We have been asked to find out what percentage of purchases are individual tracks vs whole albums, so that management can use this data to understand the effect this decision might have on overall revenue.

It is very common when you are performing an analysis to have 'edge cases' which prevent us from getting a 100% accurate answer to our question. In this instance, we have two edge cases to consider:

Albums that have only one or two tracks are likely to be purchased by customers as part of a collection of individual tracks.
Customers may decide to manually select every track from an album, and then add a few individual tracks from other albums to their purchase.

In the first case, since our analysis is concerned with maximizing revenue we can safely ignore albums consisting of only a few tracks. The company has previously done analysis to confirm that the second case does not happen often, so we can ignore this case also.

In order to answer the question, we're going to have to identify whether each invoice has all the tracks from an album. We can do this by getting the list of tracks from an invoice and comparing it to the list of tracks from an album. We can find the album to compare the purchase to by looking up the album that one of the purchased tracks belongs to. It doesn't matter which track we pick, since if it's an album purchase, that album will be the same for all tracks

We will Write a query that categorizes each invoice as either an album purchase or not, and calculates the following summary statistics:

1. Number of invoices
2. Percentage of invoices

In [ ]:
%%sql
SELECT MAX(invoice_id) total_invoice
  FROM invoice_line;

From the above table we observe that there are total of 614 invoices. Further we will compare the tracks from invoice with that of albums.

In [17]:
%%sql
WITH com_pur AS
(SELECT il.invoice_id invoice_id, il.track_id Pur_tracks, t.album_id album_id
  FROM track t
INNER JOIN invoice_line il ON il.track_id = t.track_id
ORDER BY 1)

SELECT *
  FROM com_pur;

Done.


[(1, 1158, 91),
 (1, 1159, 91),
 (1, 1160, 91),
 (1, 1161, 91),
 (1, 1162, 91),
 (1, 1163, 91),
 (1, 1164, 91),
 (1, 1165, 91),
 (1, 1166, 91),
 (1, 1167, 91),
 (1, 1168, 91),
 (1, 1169, 91),
 (1, 1170, 91),
 (1, 1171, 91),
 (1, 1172, 91),
 (1, 1173, 91),
 (2, 3476, 322),
 (2, 482, 39),
 (2, 2701, 218),
 (2, 1641, 134),
 (2, 1119, 73),
 (2, 2324, 190),
 (2, 201, 20),
 (2, 819, 66),
 (2, 392, 34),
 (2, 2651, 215),
 (3, 2516, 203),
 (3, 2646, 214),
 (4, 3448, 314),
 (4, 2560, 207),
 (4, 3336, 260),
 (4, 829, 67),
 (4, 1872, 153),
 (4, 748, 58),
 (4, 1778, 146),
 (4, 2514, 203),
 (5, 1986, 163),
 (5, 1987, 163),
 (5, 1988, 163),
 (5, 1989, 163),
 (5, 1990, 163),
 (5, 1991, 163),
 (5, 1992, 163),
 (5, 1993, 163),
 (5, 1994, 163),
 (5, 1995, 163),
 (5, 1996, 163),
 (5, 1997, 163),
 (5, 1998, 163),
 (5, 1999, 163),
 (5, 2000, 163),
 (5, 2001, 163),
 (5, 2002, 163),
 (6, 1045, 83),
 (6, 30, 5),
 (7, 3477, 322),
 (7, 3439, 305),
 (7, 1102, 87),
 (7, 711, 55),
 (7, 1988, 163),
 (7, 3024, 239),
 (7, 834, 67),
 (7, 2596, 211),
 (7, 1343, 106),
 (7, 2644, 214),
 (7, 42, 6),
 (8, 2231, 183),
 (8, 2509, 203),
 (8, 81, 9),
 (8, 1758, 145),
 (8, 1455, 118),
 (8, 2946, 233),
 (8, 574, 46),
 (8, 1752, 144),
 (8, 2263, 185),
 (8, 1560, 125),
 (9, 206, 21),
 (9, 3456, 321),
 (9, 1041, 83),
 (9, 1279, 101),
 (9, 2098, 173),
 (9, 3399, 271),
 (9, 1983, 162),
 (9, 993, 79),
 (9, 196, 20),
 (10, 2663, 215),
 (10, 3290, 257),
 (11, 2597, 211),
 (11, 2203, 181),
 (11, 2420, 197),
 (11, 2643, 214),
 (11, 3072, 243),
 (11, 1138, 89),
 (11, 2391, 195),
 (11, 2257, 185),
 (11, 610, 49),
 (11, 1419, 115),
 (11, 3452, 318),
 (12, 1463, 118),
 (12, 690, 54),
 (12, 2663, 215),
 (12, 92, 10),
 (13, 2553, 206),
 (14, 541, 42),
 (15, 3050, 241),
 (15, 1489, 120),
 (15, 807, 65),
 (15, 3336, 260),
 (16, 2160, 178),
 (16, 875, 70),
 (16, 83, 9),
 (16, 16, 4),
 (16, 1756, 145),
 (16, 3460, 321),
 (16, 93, 10),
 (16, 24, 5),
 (16, 21, 4),
 (17, 2400, 195),
 (17, 1610, 131),
 (17, 1973, 162),
 (17, 2553, 206),
 (17, 59, 7),
 (17, 2270, 185),
 (17, 55, 7),
 (17, 2564, 207),
 (17, 2171, 179),
 (17, 2522, 203),
 (17, 1419, 115),
 (18, 1027, 82),
 (18, 3142, 141),
 (18, 1583, 127),
 (18, 1752, 144),
 (18, 2736, 221),
 (19, 105, 11),
 (19, 2669, 216),
 (19, 1784, 146),
 (19, 3460, 321),
 (19, 2640, 214),
 (19, 1310, 103),
 (19, 2510, 203),
 (20, 905, 72),
 (20, 102, 11),
 (20, 3336, 260),
 (20, 1132, 88),
 (20, 337, 30),
 (20, 60, 7),
 (20, 3357, 267),
 (20, 3225, 252),
 (20, 3297, 257),
 (20, 3297, 257),
 (21, 1461, 118),
 (21, 2556, 207),
 (21, 1434, 116),
 (21, 3135, 141),
 (21, 1473, 119),
 (21, 1502, 121),
 (21, 13, 1),
 (21, 2971, 235),
 (21, 2561, 207),
 (22, 1000, 80),
 (22, 383, 33),
 (23, 1, 1),
 (23, 6, 1),
 (23, 7, 1),
 (23, 8, 1),
 (23, 9, 1),
 (23, 10, 1),
 (23, 11, 1),
 (23, 12, 1),
 (23, 13, 1),
 (23, 14, 1),
 (24, 1146, 90),
 (24, 1147, 90),
 (24, 1148, 90),
 (24, 1149, 90),
 (24, 1150, 90),
 (24, 1151, 90),
 (24, 1152, 90),
 (24, 1153, 90),
 (24, 1154, 90),
 (24, 1155, 90),
 (24, 1156, 90),
 (24, 1157, 90),
 (25, 46, 6),
 (25, 948, 75),
 (25, 2185, 180),
 (25, 3049, 241),
 (25, 1167, 91),
 (25, 32, 5),
 (25, 1800, 147),
 (25, 1477, 119),
 (26, 2670, 216),
 (26, 85, 10),
 (27, 3439, 305),
 (27, 3014, 238),
 (27, 1280, 101),
 (27, 1496, 121),
 (27, 2229, 183),
 (27, 1445, 117),
 (27, 2230, 183),
 (27, 3105, 246),
 (28, 36, 5),
 (28, 3456, 321),
 (28, 932, 74),
 (29, 1025, 82),
 (29, 2462, 200),
 (29, 3341, 261),
 (29, 1626, 132),
 (29, 2395, 195),
 (29, 479, 39),
 (30, 919, 73),
 (30, 3461, 321),
 (30, 33, 5),
 (30, 2341, 191),
 (30, 1139, 89),
 (30, 3109, 246),
 (30, 2013, 164),
 (30, 990, 79),
 (30, 2413, 196),
 (30, 2396, 195),
 (30, 3464, 321),
 (31, 2731, 221),
 (31, 2732, 221),
 (31, 2733, 221),
 (31, 2734, 221),
 (31, 2735, 221),
 (31, 2736, 221),
 (31, 2737, 221),
 (31, 2738, 221),
 (31, 2739, 221),
 (31, 2740, 221),
 (31, 2741, 221),
 (31, 2742, 221),
 (31, 2743, 221),
 (31, 2744, 221),
 (31, 2745, 221),
 (31, 2746, 221),
 (31, 2747, 221),
 (31, 27

From the above table we observe that if the tracks belonging to each distinct invoice, belongs to the same album it is an Album Purchase. If these tracks belong to different albums it is Non Album Purchase.

Below we categorize these invoices into Album and Non Album Purchase. We do this by counting the number of distinct albums belonging to each invoice. Then we compare the Distinct Albums and Number of Albums belonging to each invoice

In [18]:
%%sql
WITH com_pur AS
(SELECT il.invoice_id invoice_id, il.track_id Pur_tracks, t.album_id album_id
  FROM track t
INNER JOIN invoice_line il ON il.track_id = t.track_id
ORDER BY 1)

SELECT invoice_id, COUNT(DISTINCT(album_id)) Distinct_album_id, 
       COUNT(album_id) No_of_album_id,
       CASE
       WHEN COUNT(album_id) - COUNT(DISTINCT(album_id)) <= 2 THEN "Non Album Purchase"
       ELSE "Album Purchase"
       END AS Category 
  FROM com_pur
GROUP BY 1;

Done.


[(1, 1, 16, 'Album Purchase'),
 (2, 10, 10, 'Non Album Purchase'),
 (3, 2, 2, 'Non Album Purchase'),
 (4, 8, 8, 'Non Album Purchase'),
 (5, 1, 17, 'Album Purchase'),
 (6, 2, 2, 'Non Album Purchase'),
 (7, 11, 11, 'Non Album Purchase'),
 (8, 10, 10, 'Non Album Purchase'),
 (9, 9, 9, 'Non Album Purchase'),
 (10, 2, 2, 'Non Album Purchase'),
 (11, 11, 11, 'Non Album Purchase'),
 (12, 4, 4, 'Non Album Purchase'),
 (13, 1, 1, 'Non Album Purchase'),
 (14, 1, 1, 'Non Album Purchase'),
 (15, 4, 4, 'Non Album Purchase'),
 (16, 8, 9, 'Non Album Purchase'),
 (17, 10, 11, 'Non Album Purchase'),
 (18, 5, 5, 'Non Album Purchase'),
 (19, 7, 7, 'Non Album Purchase'),
 (20, 9, 10, 'Non Album Purchase'),
 (21, 8, 9, 'Non Album Purchase'),
 (22, 2, 2, 'Non Album Purchase'),
 (23, 1, 10, 'Album Purchase'),
 (24, 1, 12, 'Album Purchase'),
 (25, 8, 8, 'Non Album Purchase'),
 (26, 2, 2, 'Non Album Purchase'),
 (27, 7, 8, 'Non Album Purchase'),
 (28, 3, 3, 'Non Album Purchase'),
 (29, 6, 6, 'Non Album Purchase'),
 (30, 10, 11, 'Non Album Purchase'),
 (31, 1, 20, 'Album Purchase'),
 (32, 1, 13, 'Album Purchase'),
 (33, 4, 4, 'Non Album Purchase'),
 (34, 1, 1, 'Non Album Purchase'),
 (35, 5, 5, 'Non Album Purchase'),
 (36, 7, 7, 'Non Album Purchase'),
 (37, 10, 10, 'Non Album Purchase'),
 (38, 7, 7, 'Non Album Purchase'),
 (39, 8, 8, 'Non Album Purchase'),
 (40, 10, 10, 'Non Album Purchase'),
 (41, 8, 8, 'Non Album Purchase'),
 (42, 8, 9, 'Non Album Purchase'),
 (43, 7, 7, 'Non Album Purchase'),
 (44, 1, 12, 'Album Purchase'),
 (45, 8, 8, 'Non Album Purchase'),
 (46, 1, 11, 'Album Purchase'),
 (47, 6, 6, 'Non Album Purchase'),
 (48, 4, 4, 'Non Album Purchase'),
 (49, 4, 5, 'Non Album Purchase'),
 (50, 7, 7, 'Non Album Purchase'),
 (51, 8, 8, 'Non Album Purchase'),
 (52, 1, 12, 'Album Purchase'),
 (53, 3, 3, 'Non Album Purchase'),
 (54, 1, 18, 'Album Purchase'),
 (55, 9, 10, 'Non Album Purchase'),
 (56, 1, 17, 'Album Purchase'),
 (57, 1, 1, 'Non Album Purchase'),
 (58, 6, 6, 'Non Album Purchase'),
 (59, 11, 11, 'Non Album Purchase'),
 (60, 8, 8, 'Non Album Purchase'),
 (61, 1, 12, 'Album Purchase'),
 (62, 3, 3, 'Non Album Purchase'),
 (63, 5, 6, 'Non Album Purchase'),
 (64, 1, 13, 'Album Purchase'),
 (65, 2, 2, 'Non Album Purchase'),
 (66, 7, 7, 'Non Album Purchase'),
 (67, 11, 11, 'Non Album Purchase'),
 (68, 11, 11, 'Non Album Purchase'),
 (69, 1, 14, 'Album Purchase'),
 (70, 1, 1, 'Non Album Purchase'),
 (71, 1, 17, 'Album Purchase'),
 (72, 3, 3, 'Non Album Purchase'),
 (73, 1, 15, 'Album Purchase'),
 (74, 2, 2, 'Non Album Purchase'),
 (75, 4, 4, 'Non Album Purchase'),
 (76, 1, 13, 'Album Purchase'),
 (77, 1, 6, 'Album Purchase'),
 (78, 9, 9, 'Non Album Purchase'),
 (79, 1, 14, 'Album Purchase'),
 (80, 1, 1, 'Non Album Purchase'),
 (81, 6, 6, 'Non Album Purchase'),
 (82, 7, 7, 'Non Album Purchase'),
 (83, 4, 4, 'Non Album Purchase'),
 (84, 1, 12, 'Album Purchase'),
 (85, 4, 5, 'Non Album Purchase'),
 (86, 1, 1, 'Non Album Purchase'),
 (87, 8, 8, 'Non Album Purchase'),
 (88, 5, 5, 'Non Album Purchase'),
 (89, 2, 2, 'Non Album Purchase'),
 (90, 1, 11, 'Album Purchase'),
 (91, 11, 11, 'Non Album Purchase'),
 (92, 1, 20, 'Album Purchase'),
 (93, 10, 10, 'Non Album Purchase'),
 (94, 8, 8, 'Non Album Purchase'),
 (95, 11, 11, 'Non Album Purchase'),
 (96, 6, 6, 'Non Album Purchase'),
 (97, 6, 6, 'Non Album Purchase'),
 (98, 10, 10, 'Non Album Purchase'),
 (99, 1, 18, 'Album Purchase'),
 (100, 6, 6, 'Non Album Purchase'),
 (101, 4, 4, 'Non Album Purchase'),
 (102, 10, 10, 'Non Album Purchase'),
 (103, 6, 6, 'Non Album Purchase'),
 (104, 2, 2, 'Non Album Purchase'),
 (105, 7, 8, 'Non Album Purchase'),
 (106, 8, 8, 'Non Album Purchase'),
 (107, 11, 11, 'Non Album Purchase'),
 (108, 5, 5, 'Non Album Purchase'),
 (109, 11, 11, 'Non Album Purchase'),
 (110, 6, 6, 'Non Album Purchase'),
 (111, 5, 5, 'Non Album Purchase'),
 (112, 1, 17, 'Album Purchase'),
 (113, 10, 10, 'Non Album Purchase'),
 (114, 1, 17, 'Album Purchase'),
 (115, 9, 9, 'Non Album Purchase'),
 (11

In [19]:
%%sql
WITH com_pur AS
(SELECT il.invoice_id invoice_id, il.track_id Pur_tracks, t.album_id album_id
  FROM track t
INNER JOIN invoice_line il ON il.track_id = t.track_id
ORDER BY 1),

ctgr_pur AS
(SELECT invoice_id, COUNT(DISTINCT(album_id)) Distinct_album_id, 
       COUNT(album_id) No_of_album_id,
       CASE
       WHEN COUNT(album_id) - COUNT(DISTINCT(album_id)) <= 2 THEN "Non Album Purchase"
       ELSE "Album Purchase"
       END AS Category 
  FROM com_pur
GROUP BY 1)

SELECT (SELECT MAX(invoice_id)
          FROM invoice_line) TOTAL_invoice_id,
       COUNT(invoice_id) No_of_invoice_id,
       (SELECT ROUND(CAST(COUNT(invoice_id) AS FLOAT) * 100 / (SELECT MAX(invoice_id)
                                            FROM invoice_line), 2)) Perc_invoice
  FROM ctgr_pur
WHERE Category = "Non Album Purchase";

Done.


[(614, 480, 78.18)]

From the above table we deduce that out of 614 invoices, 480 invoices are Non Album Purchases. i.e. ~ 78% of purchases are Non Album Purchases.

Conclusions
In this project we answered business queries related to Chinook Record Store using SQL Database. Our analysis led to the following observations:-

We found which genres sell best in the USA. Based on this we listed three albums to be added to the Chinook store. These are:-
Red Tone
Slim Jim Bites
Meteor and the Girls

We analyzed the performance of Sales Support Agents. And found that Jane Peacock and Steve Johnson are doing better than Margaret Parker. We also observed that the age impacts the performance as Jane Peacock is youngest of all and Steve Johnson is younger to Margaret Parker.
We analyzed the sales data for customers from each country. We calculated total number of customers, total value of sales, average value of sales per customer and average order value for each country. We observed that USA is on top of the list with 13 customers and ~ $1040 total value of sales. Next is CANADA with 8 customers and with ~ $535 value of sales.

We calculated the percentage of individual tracks and the percentage of whole album purchased by customers. Our analysis led to the conclusion that the individual tracks purchases (~ 78%) are higher compared to album purchases (~ 22%). This can help the company in deciding whether to buy popular tracks from albums or to buy the whole album from record companies.


